Mathlib/Algebra/Module/NestAlgebra.lean

```diff
@@ -0,0 +1,91 @@
+/-
+Copyright (c) 2024 Christopher Hoskin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christopher Hoskin
+-/
+
+import Mathlib.Algebra.Module.Submodule.Lattice
+import Mathlib.Order.Sublattice
+import Mathlib.Algebra.Algebra.Subalgebra.Basic
+import Mathlib.LinearAlgebra.Basis.Defs
+
+/-!
+# Nest algebras
+
+Nest algebras are an operator analogue of upper triangular matrices. They have been extensively
+studied in the context of Hilbert spaces, but also in more general contexts.
+
+This file develops the algebraic aspects of the theory.
+
+## Main definitions
+
+- `NestAlg`: The nest algebra generated by a `Nest`
+- `Basis.toNest`: Every basis of a free module generates a `Nest` (and hence a nest subalgebra)
+
+## References
+
+- [K. R. Davidson, *Nest algebras*][davidson1988]
+- [Z. Mesyan, *Infinite-Dimensional Triangularization*][mesyan2018]
+
+## Tags
+
+nest algebra, triangularization
+-/
+
+variable {R M : Type*}
+
+variable [CommSemiring R] [AddCommMonoid M] [Module R M]
+
+/--
+The nest algebra of a nest
+-/
+def NestAlg (N : Sublattice (Submodule R M)) : Subalgebra R (M →ₗ[R] M) where
+  carrier := { T | ∀ (n : N), T '' n ⊆ n}
```

Comment:
1. There's nothing in this definition that needs `Nest` or even `Sublattice (Submodule R M)`.
2. Note that `T '' n \subset n` can now be phrased `n \in T.invtSubmodule`.
3. You can define, for a single submodule `p`, the subalgebra with carrier `{T | p \in T.invtSubmodule}`.
4. Then you can recover this subalgebra as the `\Inf` over all submodules in a given set of submodules, it need not be a lattice or a nest.

Reply:
Thanks.
```
def NestAlg (N : Set (Submodule R M)) : Subalgebra R (M →ₗ[R] M) where
  carrier := { T | ∀ (n : N), ↑n ∈ Module.End.invtSubmodule (T : Module.End R M)}
```
also seems to work - but perhaps there is some advantage to defining the subalgebra generated by a single submodule first?